### DISCUSSION TITLE
Add Event hooks to Plugins

### DISCUSSION THREAD

**Discussion by CEbbinghaus** [Kong/insomnia#3692] [Ideas]
**Is your feature request related to a problem? Please describe.**
There aren't any ways to initialize plugins. Reading Settings and other tasks have to be done in the actions that they are relevant for.

**Describe the solution you'd like**
Methods to handle LifeCycle hooks of the Plugin such as `OnLoad`, `OnUnload`, `OnReload`...etc, as well as event callbacks of the Application which could include: `OnWorkspaceOpen` or more granular events: `OnDocumentOpen`, `OnCollectionOpen`

**Describe alternatives you've considered**
Running code when the file is executed is easy and certainly possible but it lacks access to the Insomnia plugin context. e.g no saving to the store to read later or alerting the user to changes in a new version.

**Additional context**
Related to #2599 which would be another useful hook

**Comment by dimitropoulos** [comment]
Are there specific events you have in mind in the application that you'd like to capture that _are not_ redux actions?

**Comment by CEbbinghaus** [comment]
> Are there specific events you have in mind in the application that you'd like to capture that _are not_ redux actions?

well I specified a few in the description but some I would like to see are

**Plugin Events:**
* `OnLoad` *(Called after a plugin is loaded. Could be split up into `OnLoad` (one time) and `Initialize` (every load))*
* `OnUnload` *(Called before a plugin is Unloaded. Useful for plugins to ensure their data is saved)*
* `OnReload` *(Called before a plugin is Reloaded. Could be skipped since it would also call `OnUnload`, but always useful to separate the two)*
* `OnUpdate` *(Called after a plugin Updates, e,g to update configs or alert the user to a change #2599)*

**Insomnia Events:**
* `OnWorkspaceOpen` *(whenever the user opens a document or collection)*
    * `OnDocumentOpen` *(more specific event for only documents)*
    * `OnCollectionOpen` *(and Collections)*
 * `OnApplicationClose` *(before the app closes. e.g so any config changes can be written to disk )*
 * `OnWorkspaceUpdate` *(When a change is pulled via Sync or a workspace file is imported)*

**Comment by dimitropoulos** [comment]
Wow great! That's a very very helpful and clear list.  cc: @wdawson
